### PR TITLE
[zh-cn]: Update translation of String.matchAll()

### PR DESCRIPTION
**Pull request by JasonLamv-t**
### Description

Same as the title.

### Motivation

Ditto.

### Additional details

None.

### Related issues and pull requests

#13975

### CODE REVIEW

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 1]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
-regexp
+- `regexp`
 
-正则表达式对象。如果所传参数不是一个正则表达式对象，则会隐式地使用 `new RegExp(obj)` 将其转换为一个 [`RegExp`](/zh-CN/docs/Web/JavaScript/Reference/RegExp)。
+  - : 一个正则表达式对象，或者是任何具有 [`Symbol.matchAll`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/matchAll) 方法的对象。
 
-`RegExp`必须是设置了全局模式`g`的形式，否则会抛出异常`TypeError`。
+    如果 `regexp` 不是一个 `RegExp` 对象，并且没有 `Symbol.matchAll` 方法，它将通过 `new RegExp(regexp, 'g')` 被隐式转换为一个 {{jsxref("RegExp")}} 对象。
+
+    如果 `regexp` 是一个[正则表达式对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，那么它必须设置了全局（`g`） 标志，否则会抛出一个 {{jsxref("TypeError")}} 异常。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:26]: extra space

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 2]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
 
 ### 返回值
 
-一个迭代器（不可重用，结果耗尽需要再次调用方法，获取一个新的迭代器）。
+一个匹配结果的[可迭代的迭代器对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Iterator)（它不可重新开始）。每个匹配结果都是一个数组，其形状与 {{jsxref("RegExp.prototype.exec()")}} 的返回值相同。
+
+### 异常
+
+- {{jsxref("TypeError")}}
+  - : 如果 `regexp` 是一个[正则表达式对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，且没有设置全局（`g`） 标志（其 [`flags`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp/flags) 属性不包含 `"g"`），则会抛出该异常。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:35]: same

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md (hunk 1 above)
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:26]:
```suggestion
    如果 `regexp` 是一个[正则表达式对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，那么它必须设置了全局（`g`）标志，否则会抛出一个 {{jsxref("TypeError")}} 异常。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md (hunk 2 above)
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:35]:
```suggestion
  - : 如果 `regexp` 是一个[正则表达式对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，且没有设置全局（`g`）标志（其 [`flags`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp/flags) 属性不包含 `"g"`），则会抛出该异常。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 3]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
 
 {{JSRef}}
 
-**`matchAll()`** 方法返回一个包含所有匹配正则表达式的结果及分组捕获组的迭代器。
+**`matchAll()`** 方法返回一个迭代器，用于匹配一个字符串和一个[正则表达式](/zh-CN/docs/Web/JavaScript/Guide/Regular_expressions)，包括[捕获组](/zh-CN/docs/Web/JavaScript/Guide/Regular_expressions/Groups_and_backreferences)在内的所有结果。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:8]: 翻译再改改

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 4]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
-regexp
+- `regexp`
 
-正则表达式对象。如果所传参数不是一个正则表达式对象，则会隐式地使用 `new RegExp(obj)` 将其转换为一个 [`RegExp`](/zh-CN/docs/Web/JavaScript/Reference/RegExp)。
+  - : 一个正则表达式对象，或者是任何具有 [`Symbol.matchAll`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/matchAll) 方法的对象。
 
-`RegExp`必须是设置了全局模式`g`的形式，否则会抛出异常`TypeError`。
+    如果 `regexp` 不是一个 `RegExp` 对象，并且没有 `Symbol.matchAll` 方法，它将通过 `new RegExp(regexp, 'g')` 被隐式转换为一个 {{jsxref("RegExp")}} 对象。
+
+    如果 `regexp` 是一个[正则表达式对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，那么它必须设置了全局（`g`）标志，否则会抛出一个 {{jsxref("TypeError")}} 异常。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:26]:
```suggestion
    如果 `regexp` 是一个[正则表达式](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，那么它必须设置了全局（`g`）标志，否则会抛出 {{jsxref("TypeError")}} 异常。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 5]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
 
 ### 返回值
 
-一个迭代器（不可重用，结果耗尽需要再次调用方法，获取一个新的迭代器）。
+一个匹配结果的[可迭代的迭代器对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Iterator)（它不可重新开始）。每个匹配结果都是一个数组，其形状与 {{jsxref("RegExp.prototype.exec()")}} 的返回值相同。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:30]:
```suggestion
一个匹配结果的[可迭代迭代器对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Iterator)（它不可重新开始）。每个匹配结果都是一个数组，其形状与 {{jsxref("RegExp.prototype.exec()")}} 的返回值相同。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 6]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
+
+### 异常
+
+- {{jsxref("TypeError")}}
+  - : 如果 `regexp` 是一个[正则表达式对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，且没有设置全局（`g`）标志（其 [`flags`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp/flags) 属性不包含 `"g"`），则会抛出该异常。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:35]:
```suggestion
  - : 如果 `regexp` 是一个[正则表达式](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp#正则表达式的特殊处理)，且没有设置全局（`g`）标志（其 [`flags`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp/flags) 属性不包含 `"g"`），则会抛出该异常。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 7]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
+  console.log(
+    `找到 ${match[0]} 起始位置=${match.index} 结束位置=${regexp.lastIndex}。`,
+  );
 }
+// 找到 football 起始位置=6 结束位置=14。
+// 找到 foosball 起始位置=16 结束位置=24。
 ```
 
-如果使用 `matchAll` ，就可以不必使用 while 循环加 exec 方式（且正则表达式需使用 `/g` 标志）。使用 `matchAll` 会得到一个迭代器的返回值，配合 [`for...of`](/zh-CN/docs/Web/JavaScript/Reference/Statements/for...of), [array spread](/zh-CN/docs/Web/JavaScript/Reference/Operators/Spread_syntax), 或者 {{jsxref("Array.from()")}} 可以更方便实现功能：
+有了可用的 `matchAll()` 方法，可以避免使用 `while` 循环和带有 `g` 标志的 `exec`。相反，你可以获得一个迭代器，以便使用更方便的 `for...of` 循环、[数组展开语法](/zh-CN/docs/Web/JavaScript/Reference/Operators/Spread_syntax)或者 `Array.from()` 构造函数：
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:61]:
```suggestion
如果有 `matchAll()` 方法，则可以避免使用 {{jsxref("Statements/while", "while")}} 循环和带有 `g` 标志的 `exec`。相反，你可以获得一个迭代器，以便使用更方便的 {{jsxref("Statements/for...of", "for...of")}} 循环、[数组展开语法](/zh-CN/docs/Web/JavaScript/Reference/Operators/Spread_syntax)或 `Array.from()` 构造函数：
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 8]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
-// Call matchAll again to create a new iterator
-Array.from(str.matchAll(regexp), m => m[0]);
-// Array [ "football", "foosball" ]
+// 匹配迭代器在 for...of 迭代后用尽
+// 再次调用 matchAll 以创建新的迭代器
+Array.from(str.matchAll(regexp), (m) => m[0]);
+// [ "football", "foosball" ]
 ```
 
 如果没有 `/g` 标志，`matchAll` 会抛出异常。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:84]:
```suggestion
如果没有 `g` 标志，`matchAll` 会抛出异常。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 9]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
-Array.from(str.matchAll(regexp), m => m[0]);
-// TypeError: String.prototype.matchAll called with a non-global RegExp argument
+const regexp = /[a-c]/;
+const str = "abc";
+str.matchAll(regexp);
+// TypeError
 ```
 
-`matchAll` 内部做了一个 regexp 的复制，所以不像 [regexp.exec](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec), `lastIndex` 在字符串扫描时不会改变。
+`matchAll` 内部做了一个 `regexp` 的复制，所以不像 {{jsxref("Global_Objects/RegExp/exec", "regexp.exec()")}}，`lastIndex` 在字符串扫描时不会改变。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:93]:
```suggestion
`matchAll` 内部做了一个 `regexp` 的复制，所以不像 {{jsxref("Global_Objects/RegExp/exec", "regexp.exec()")}}，`lastIndex` 在扫描字符串后不会改变。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 10]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
+Array.from(str.matchAll(regexp), (m) => `${regexp.lastIndex} ${m[0]}`);
+// [ "1 b", "1 c" ]
 ```
 
-### 捕获组的更佳途径
+然而，这也意味着，与在循环中使用 [`regexp.exec()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec) 不同，你不能更改 `lastIndex` 来使正则表达式前进或倒退。
+
+### 比 String.prototype.match() 更好的捕获组获取方式
+
+`matchAll()` 方法的另一个重要优点是改进了对于捕获组的获取方式。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:107]: 请将“方法名加括号”的修改同步至英文文档

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 11]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
+const regexp = /t(e)(st(\d?))/g;
+const str = "test1test2";
 
-str.match(regexp);
-// Array ['test1', 'test2']
+str.match(regexp); // ['test1', 'test2']
 ```
 
-使用 `matchAll` 可以通过如下方式获取分组捕获：
+使用 `matchAll()` 方法，可以轻松获取捕获组：
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:118]: 包括下面的修改

File: files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md [hunk 12]
```diff
@@ -5,107 +5,141 @@ slug: Web/JavaScript/Reference/Global_Objects/String/matchAll
...
-let array = [...str.matchAll(regexp)];
+const array = [...str.matchAll(regexp)];
 
 array[0];
 // ['test1', 'e', 'st1', '1', index: 0, input: 'test1test2', length: 4]
 array[1];
 // ['test2', 'e', 'st2', '2', index: 5, input: 'test1test2', length: 4]
 ```
 
+### 使用一个没有实现 @@matchAll 的非正则对象调用 matchAll()
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/matchall/index.md:129]:
```suggestion
### 使用实现了 @@matchAll 的非正则对象调用 matchAll()
```

**Review by yin1999 — CHANGES_REQUESTED** [review]

**Review by yin1999 — APPROVED** [review]
Thanks. Looks better 😉

**Pull request merged**